src/firmware_new/boards/frankie_v1/udp_multicast.c

```diff
@@ -0,0 +1,117 @@
+#include "firmware_new/proto/control_fw.pb.h"
+#include "lwip/api.h"
+#include "lwip/inet.h"
+#include "lwip/ip_addr.h"
+#include "lwip/memp.h"
+#include "lwip/opt.h"
+#include "lwip/pbuf.h"
+#include "lwip/sys.h"
+#include "lwip/tcp.h"
+#include "lwip/udp.h"
+#include "main.h"
+#include "pb.h"
+#include "pb_decode.h"
+#include "pb_encode.h"
+
+// TODO the messages are global for now, will need to be moved
+// when this file is properly integrated. This file only serves
+// as a reference implementation of reliable multicast with IPv6
+robot_ack ack       = robot_ack_init_zero;
+control_msg control = control_msg_init_zero;
+
+// this buffer is used to hold serialized proto. Allocated
+// on the data segment to avoid mallocing on every packet received
+static uint8_t buffer[robot_ack_size];
+
+// the thread that handles multicast recv/send
+static void udp_multicast_thread(void *arg);
+
+struct multicast_config
+{
+    ip_addr_t multicast_address;
+    unsigned multicast_port;
+    unsigned send_port;
+} typedef multicast_config_t;
+
+/*
+ * Thread that creates a send and recv socket, joins the specified
+ * multicast group and start listening for packets. Increments a counter
+ * on every multicast packet and sends a unicast packet back to the sender.
+ *
+ */
+static void udp_multicast_thread(void *arg)
+{
+    multicast_config_t *config = (multicast_config_t *)arg;
```

Comment:
Any reason we can't just pass a `multicast_config_t*` as the parameter rather than a `void*`?

Reply:
nope no reason, I just saw the function signature took a void ptr, I guess it can be implicitly casted

Reply:
I guess just to be clear, we control this function signature right? So we can just change the parameter to accept a `multicast_config_t*`?

Reply:
Right, but I just tested this out and I get 

```
firmware_new/boards/frankie_v1/udp_multicast.c:113:40: error: passing argument 2 of 'sys_thread_new' from incompatible pointer type [-Werror=incompatible-pointer-types]
     sys_thread_new("multicast_thread", blocking_udp_multicast_loop, (multicast_config_t*)config, 1024,
```

The `sys_thread_new` signature requires a void ptr for the thread

Reply:
Yeah, this makes sense.

Reply:
ah ok